internal/overlord/logstate/syslog.go

```diff
@@ -0,0 +1,184 @@
+package logstate
+
+import (
+	"bytes"
+	"fmt"
+	"net"
+	"net/url"
+	"sync"
+	"time"
+
+	"github.com/canonical/pebble/internal/plan"
+)
+
+func init() {
+	RegisterLogBackend("syslog", func(t *plan.LogTarget) (LogBackend, error) { return NewSyslogBackend(t.Location) })
+}
+
+func validateSyslog(t *plan.LogTarget) error {
+	b, err := NewSyslogBackend(t.Location)
+	if b != nil {
+		b.Close()
+	}
+	return err
+}
+
+// SyslogBackend takes writes and formats them according to RFC5424.  The formatted syslog messages
+// are then forwarded to the underlying syslog server.  SyslogWriter is safe for concurrent writes
+// and use.
+type SyslogBackend struct {
+	version        int
+	host           string
+	msgid          string
+	priority       int
+	pid            string
+	structuredData string // TODO: use this for future log labels
+
+	// mu helps manage access to network connection-related internal state.  This should be
+	// read-locked whenever using the syslog server connection and write-locked whenever
+	// replacing/reconnecting/etc the syslog server connection.
+	mu            sync.RWMutex
+	conn          net.Conn
+	address       *url.URL
+	waitReconnect time.Duration
+	closed        bool
+}
+
+// NewSyslogBackend creates a writer forwarding writes as syslog messages to dst.  The forwarded
+// messages will have app as the application name.  Other message parameters are set using
+// reasonable defaults or the RFC5424 nil value "-".
+func NewSyslogBackend(addr string) (*SyslogBackend, error) {
+	u, err := url.Parse(addr)
+	if err != nil || u.Host == "" {
+		u, err = url.Parse("//" + addr)
+	}
+	if err != nil {
+		return nil, err
+	} else if u.Scheme != "tcp" && u.Scheme != "udp" && u.Scheme != "" {
+		return nil, fmt.Errorf("invalid syslog server address scheme %q", u.Scheme)
+	}
+	if u.Scheme == "" {
+		u.Scheme = "tcp"
+	}
+
+	return &SyslogBackend{
+		version:        1,
+		pid:            "-",
+		host:           "-",
+		msgid:          "-",
+		priority:       1*8 + 6, // for facility=user-msg severity=informational. See RFC 5424 6.2.1 for available codes.
```

Comment:
Can we make this a constant with docs and a good name ?

Reply:
!done